docs/DesignDocs/SGX_EMM.md

```diff
@@ -0,0 +1,678 @@
+
+
+## User Experience ##
+
+**Allocate, Deallocate Enclave Memory**
+
+The EMM provides an API, sgx_mm_alloc, for clients to request enclave memory
+allocations. An enclave memory allocation represents both a reserved virtual
+address range and a commitment of EPC pages.  EPC pages are committed for
+enclaves via special SGX instructions: loaded by EADD/EEXTEND before EINIT
+or dynamically added using EAUG followed by EACCEPT.
+
+The sgx_mm_alloc API allows clients to specify one of three committing modes
+for an allocation:
+- SGX_EMA_RESERVE, only the virtual address range is reserved. No EPC pages will
+be committed in this mode.
+- SGX_EMA_COMMIT_NOW: reserves and commits physical EPC upon allocation.
+EACCEPT will be done immediately.
+- SGX_EMA_COMMIT_ON_DEMAND: EACCEPT is done on demand, see below on committing
+and uncommitting.
+
+An allocation, once created, will own its address range until the deallocation
+API, sgx_mm_dealloc, is called upon. No two active allocations can have
+overlapping address ranges.
+
+**Commit, Uncommit Enclave Memory**
+
+When a page in COMMIT_ON_DEMAND allocations is accessed, a page fault occurs if
+the page was not yet committed.  The EMM will perform EACCEPT to commit the EPC
+page on page fault after OS doing EAUG.
+
+The clients can also call the EMM commit API, sgx_mm_commit, to proactively
+commit specific sub-regions in a COMMIT_ON_DEMAND allocation to avoid
+future page fault.
+
+Some EMM clients, <i>e.g.</i>, a dynamic code loader wishing to load code on
+page faults, can register a custom handler for page faults at the time of
+allocation request. In the custom page fault handler, it can invoke an API,
+sgx_mm_commit_data, to commit and load data to newly committed EPC page at
+the same time as supported by EACCEPTCOPY.
+
+Committed pages will stay committed (regardless how they were committed) until
+the clients calls the uncommit API, sgx_mm_uncommit, on them or the allocation
+they belong to is deallocated by sgx_mm_dealloc.
+
+**Modify Page Attributes**
+
+The EMM clients may call sgx_mm_modify/sgx_mm_modify_ex to request permissions
+and/or page type changes for pages in existing allocations.
+
+**Runtime Abstraction Layer**
+
+To make the EMM implementation portable across different SGX runtimes,
+e.g., Open Enclave and Intel SGX SDKs, this document also proposes an abstraction
+layer for the runtimes. The runtime abstraction layer APIs are invoked by the EMM
+to make necessary OCalls, register and receive callbacks on page faults. The EMM
+source code will be hosted and maintained in the [Intel SGX PSW and SDK repository](https://github.com/intel/linux-sgx)
+and can be built as a separate library to be linked into any runtime with the
+abstraction layer APIs implemented.
+
+## Notes on Internal Design ##
+
+The enclave memory manager keeps track of memory allocation and layout info inside
+enclave address range (ELRANGE) using an internal structure called the Enclave Memory
```

Comment:
Does this mean the enclave need to over request the ELRANGE during the creation?
Something like (where `ADDITIONAL REGION` is for the enclave to dynamically allocate via EMM):
```
[ENCLAVE LAYOUT][ADDITIONAL REGION]
<----          ELRANGE        ---->
```

And when 0-base enclave is enabled, the layout becomes

```
[REGION from ADDR 0][ENCLAVE LAYOUT][ADDITIONAL REGION]
<----                    ELRANGE                  ---->
```
I assume the EMM internally will ensure that `REGION from ADDR 0` should not be used?

Reply:
Runtime can return false from sgx_is_within_enclave or ema_alloc a RESERVED region for REGION from ADDR 0 with EMA_SYSTEM flag so no one else can allocate from it.